### PR TITLE
Improve annotation reader to traverse class hierarchy

### LINKED ISSUE #67: False positive for inherited descriptive annotations
Inherited description annotations are incorrectly reported as unused:

```php
interface ClassInterface {
   /*
    * @throws \LogicException Description.
    */
    public function foo()
}

class ChildClass implements ClassInterface {
    public function foo(){
        throw new \LogicException(); // Unused @throws LogicException
    } 
}
```

### PR DESCRIPTION
**Pull request by marcospassos**
Fixes #67

### CODE REVIEW

**Comment by marcospassos** [comment]
@pepakriz can you review it?

File: src/ThrowsAnnotationReader.php [hunk 1]
```diff
@@ -130,9 +130,14 @@ private function getDocblock($reflection): ?string
...
 			$declaringClass = $reflection->getDeclaringClass();
-			$nativeClassReflection = $declaringClass->getNativeReflection();
-			$nativeMethodReflection = $nativeClassReflection->getMethod($reflection->getName());
-			$docBlock = $nativeMethodReflection->getDocComment();
+			$classReflection = $declaringClass->getNativeReflection();
+			$methodReflection = $classReflection->getMethod($reflection->getName());
+			$docBlock = $methodReflection->getDocComment();
+
+			while ($docBlock === false) {
+				$methodReflection = $methodReflection->getPrototype();
```
Comment (pepakriz) [inline src/ThrowsAnnotationReader.php]: It's throwing an exception when a prototype is not there. It should return `null` instead.

File: src/ThrowsAnnotationReader.php [hunk 2]
```diff
@@ -130,9 +130,14 @@ private function getDocblock($reflection): ?string
...
+				$docBlock = $methodReflection->getDocComment();
+			}
 
 			return $docBlock !== false ? $docBlock : null;
```
Comment (pepakriz) [inline src/ThrowsAnnotationReader.php:147]: `$docBlock` can't be false in this case.

**Review by pepakriz — CHANGES_REQUESTED** [review]

File: src/ThrowsAnnotationReader.php (hunk 1 above)
Comment (marcospassos) [reply src/ThrowsAnnotationReader.php]: Yeah, but it's [wrapped in a try-catch block](https://github.com/pepakriz/phpstan-exception-rules/blob/5f30e824dd24cb680a12c365f4244a73bff21b44/src/ThrowsAnnotationReader.php#L97).

File: src/ThrowsAnnotationReader.php (hunk 2 above)
Comment (marcospassos) [reply src/ThrowsAnnotationReader.php:147]: old code, I will fix it

File: src/ThrowsAnnotationReader.php (hunk 1 above)
Comment (marcospassos) [reply src/ThrowsAnnotationReader.php]: (except by a try-catch block, there is no way to check whether a method has a prototype)

File: src/ThrowsAnnotationReader.php (hunk 2 above)
Comment (marcospassos) [reply src/ThrowsAnnotationReader.php:147]: I can't remove it, any suggestion?
```
 ------ ---------------------------------------------------------------------- 
  Line   src/ThrowsAnnotationReader.php                                        
 ------ ---------------------------------------------------------------------- 
  147    Method                                                                
         Pepakriz\PHPStanExceptionRules\ThrowsAnnotationReader::getDocblock()  
         should return string|null but returns string|false.                   
 ------ ---------------------------------------------------------------------- 
```

File: src/ThrowsAnnotationReader.php (hunk 1 above)
Comment (marcospassos) [reply src/ThrowsAnnotationReader.php]: besides the previous comments:
```
 ------ ---------------------------------------------------------------------- 
  Line   src/ThrowsAnnotationReader.php                                        
 ------ ---------------------------------------------------------------------- 
  140    ReflectionException is never thrown in the corresponding try block    
 ------ ---------------------------------------------------------------------- 
```

Comment (pepakriz) [reply src/ThrowsAnnotationReader.php]: I know, but it's weird in my opinion. `null` value means there is no phpdoc. Thrown exception is for cases when reflection fails.

Comment (marcospassos) [reply src/ThrowsAnnotationReader.php]: done

**Comment by marcospassos** [comment]
@pepakriz is there something pending here?

File: src/ThrowsAnnotationReader.php (hunk 2 above)
Comment (pepakriz) [reply src/ThrowsAnnotationReader.php:147]: Hmm, it's a bug in phpstan :( https://phpstan.org/r/0935e908-80f9-44ec-8de8-de60f11f0a8c

So ok, keep it in this way

**Review by pepakriz — APPROVED** [review]

**Pull request merged**